### ISSUE TITLE
FreeU support

### ISSUE THREAD

**Issue by saintvaseline** [leejet/stable-diffusion.cpp#132]
FreeU is a method that seems to give somewhat more impressive results which can be used with any sampling method and SD model. A lot of artifacts are fixed, text is rendered somewhat better and the colors are somewhat less murky.

I imagine this would help counteract the "murkiness" of SDXL images' colors (SDXL turbo looks alright using euler_a, but becomes murkier again with lcm sampler). LCM images seem to be especially murky as well.

There are 2 versions of FreeU in ComfyUI. FreeU_v2, which is a newer version of the method seems to be the best. Here is a comparison: https://www.reddit.com/r/comfyui/comments/17l5spc/comparison_of_freeu_v2_with_old_freeu/

Paper: https://arxiv.org/abs/2309.11497
Repository: https://github.com/ChenyangSi/FreeU
Project Page: https://chenyangsi.top/FreeU/
Demo: https://huggingface.co/spaces/ChenyangSi/FreeU

**Comment by saintvaseline** [comment]
FreeU_v2 seems to be a paper update: https://twitter.com/scy994/status/1714499568573039102

**Comment by FSSRepo** [comment]
Interesting, I'll see if I can give it some time to at least try if it works.

**Comment by bssrdf** [comment]
I gave a try but I am not sure if implemented correctly. Here are the comparison without and with FreeU.

FreeU OFF | FreeU ON
--- | ---
![freeu-off_5](https://github.com/leejet/stable-diffusion.cpp/assets/689043/0c094d46-4c0d-487e-ab3a-b0fd61d5eb50) | ![freeu-on_5](https://github.com/leejet/stable-diffusion.cpp/assets/689043/05e516f6-4dd0-46da-b53f-cf18a0323c96) 
![freeu-off_4](https://github.com/leejet/stable-diffusion.cpp/assets/689043/065a4633-6eae-429b-b6f4-f8bcf5160506) | ![freeu-on_4](https://github.com/leejet/stable-diffusion.cpp/assets/689043/d7ffe9d0-b30c-400b-8159-035e167e4775)  
![freeu-off_3](https://github.com/leejet/stable-diffusion.cpp/assets/689043/3f659da5-5d4a-41f2-b23c-0b621edb20b2) | ![freeu-on_3](https://github.com/leejet/stable-diffusion.cpp/assets/689043/6ea1ed0e-76e3-45c7-959c-857cad56d75b)
![freeu-off_2](https://github.com/leejet/stable-diffusion.cpp/assets/689043/63d2c65a-0419-403c-9bdf-ac18601b0681) | ![freeu-on_2](https://github.com/leejet/stable-diffusion.cpp/assets/689043/7bac5c75-3520-4aef-8f01-76782edff628)

**Comment by FSSRepo** [comment]
>I gave a try but I am not sure if implemented correctly. 

@bssrdf Do you mean in sd.cpp? I checked your fork, and I don't see anything related.

**Comment by bssrdf** [comment]
> > I gave a try but I am not sure if implemented correctly.
> 
> @bssrdf Do you mean in sd.cpp? I checked your fork, and I don't see anything related.

Yes. Please check out [add-freeu-support branch](https://github.com/bssrdf/stable-diffusion.cpp/tree/add-freeu-support) of my sd.cpp fork as well as [add-freeu-support branch](https://github.com/bssrdf/ggml/tree/add-freeu-support) of my ggml fork. Most of the work is done in `ggml`. Sorry, code is a bit messy now and I will do a clean up before the PR.

**Comment by FSSRepo** [comment]
@bssrdf Great job, to implement that functionality in sd.cpp for obvious reasons, it must also be compatible with the CPU backend. I don't know much about how to implement the FFT optimally, so I can't help you with that. I'm also interested to see if you could replicate those results with sd-webui or with the implementation from the repository to check if they match or are similar. Additionally, it would be helpful to determine if certain values passed to the Fourier_filter function are the same as those you receive with your implementation.

**Comment by leejet** [comment]
> > > I gave a try but I am not sure if implemented correctly.
> > 
> > 
> > @bssrdf Do you mean in sd.cpp? I checked your fork, and I don't see anything related.
> 
> Yes. Please check out [add-freeu-support branch](https://github.com/bssrdf/stable-diffusion.cpp/tree/add-freeu-support) of my sd.cpp fork as well as [add-freeu-support branch](https://github.com/bssrdf/ggml/tree/add-freeu-support) of my ggml fork. Most of the work is done in `ggml`. Sorry, code is a bit messy now and I will do a clean up before the PR.

Great job! Can you create a PR for both ggml and sd.cpp?

**Comment by bssrdf** [comment]
@FSSRepo, @leejet, thanks. There is still quite some work before I can make a PR. 

-  The changes to `sd.cpp` are relatively trivial but they do require some interface mods (I think it is better to make a switch for FreeU on or off). 
- The `ggml` mods are more substantial, but still manageable. The problem is the two operators (FFT_FILTER and FREEU_BACKBONE) I added for FreeU are too specific and may not get accepted.  I see there are some ggml extensions in sd.cpp but I don't see how cuda backends can be added.  I am seeking your advice.  
- The two additional operators are only implemented for GPU backend and I have to work on their CPU counterparts.

@FSSRepo, I can verify that the two operators are functioning correctly as I have test cases for both. But I don't think I can replicate results with sd-webui or with the implementation from the original author's repository.
